fix(docs): add a page describing everything about feature statuses (#31941)
mdn/content: files/en-us/mdn/writing_guidelines/page_structures/banners_and_notices/index.md

> {{MDNSidebar}}

Reviewer: FYI @OnkarRuikar , I updated this to my taste in https://github.com/mdn/content/pull/31941/commits/07f16a4ba9dceeca29d619ce1c304c5f89c9da69

- Changed from Bluetooth to AudioDecoder, since that actually does require a secure context
- Added the AvailableInWorkersMacro
- Moved the discussion of the frontmatter status into a note where it is relevant.
- Removed link to Page type templates - I guess cross links are useful, but not actually all that useful in this case.
- All headings at same level. The first topic is now "Whare are banner macros added" - it's clearly just showing a single real-world example.

Hope this makes sense to you.

Author: Thank you @hamishwillee for the changes!

> Changed from Bluetooth to AudioDecoder, since that actually does require a secure context

We are in process of automating the secure context macros so those macros are also going to be in automated macros' list.

Reviewer: Awesome. That will be good. Then this can almost disappear.

For example, banners are used to highlight when a particular interface, method or property is deprecated, and should not be used in production code, or can only be used in a secure context.

Banners are rendered using macros in the page content.
Some banner macros are added automatically to the page, while others are added manually.

This article describes the more important banners and how they are added.

## Where are banner macros added

Banners are added using macros that are usually inserted below the page metadata, alongside the page sidebar macro. For example, in the block below, the `\{{SecureContext_Header}}` macro has been used to indicate that the {{domxref("AudioDecoder")}} interface is only available in a [secure context](/en-US/docs/Web/Security/Secure_Contexts), and `\{{SeeCompatTable}}` has been added to indicate that the interface is experimental.

```md
---
title: AudioDecoder
slug: Web/API/AudioDecoder
page-type: web-api-interface
status:
  - experimental
browser-compat: api.AudioDecoder
---

\{{APIRef("WebCodecs API")}}\{{SeeCompatTable}}\{{SecureContext_Header}}
```

> While most banners are inserted immediately below the page metadata, the `\{{AvailableInWorkers}}` macro is added at the bottom of the first section of text, before the first heading.

Author: Why the strange requirement? Most pages add `{{AvailableInWorkers}}` directly below page metadata. Also, some pages don't added it immediately after first para, e.g. `Web/API/CookieStoreManager` and `Web/API/Headers`.
Some pages have it above `{{InheritanceDiagram}}` and some below the InheritanceDiagram macro.

It would be easy to for the secure context automation if we have a fix location. I suggest it should be directly below the front-matter section like other status banners.

> While most banners are inserted immediately below the page metadata, the `\{{AvailableInWorkers}}` macro is added at the bottom of the first section of text, before the first heading.

Reviewer: Because the strange requirement is what I was told to use when I started here, and I expect is the most common pattern in existing docs.

I've always felt the inconsistency odd. If you want to change it then cool, but we should get broad agreement and change it everywhere.

> While most banners are inserted immediately below the page metadata, the `\{{AvailableInWorkers}}` macro is added at the bottom of the first section of text, before the first heading.

Author: We can take this up in a separate PR or a discussion.

{{MDNSidebar}}

Banners are displayed in some pages, in particular API reference, in order to highlight important factors that will affect how the described content is used.
For example, banners are used to highlight when a particular interface, method or property is deprecated, and should not be used in production code, or can only be used in a secure context.

Banners are rendered using macros in the page content.
Some banner macros are added automatically to the page, while others are added manually.

This article describes the more important banners and how they are added.

> ## How to add a banner

Reviewer: @OnkarRuikar Changes are better, but still not quite right. So this topic is "Banners and notices" but it says nothing about notices. Should it? Assuming yes, we should make this topic level 2 and put the others under it.

Would you like me to take a shot at updating this as I would like it? 

```suggestion
## Banners
```

Author: By notices do you mean `> **Warning:**` and `> **Note:**` at the top of the pages?

> Would you like me to take a shot at updating this as I would like it?

Sure. I've given you access to my fork. Run `gh pr checkout 31941` to get the PR branch in your local.

Reviewer: > By notices do you mean > **Warning:** and > **Note:** at the top of the pages?

Actually, I meant things like `{{Compat}}` which are mentioned in the section on statuses. They aren't a banner or a notice.

Banners are added using macros. Banner macros should be inserted below the page metadata, alongside the page sidebar macro. For example, in the block below, the `\{{SecureContext_Header}}` macro has been used to indicate that the {{domxref("Bluetooth")}} interface is only available in a [secure context](/en-US/docs/Web/Security/Secure_Contexts).

```md
---
title: Bluetooth
slug: Web/API/Bluetooth
page-type: web-api-interface
status:
  - experimental
browser-compat: api.Bluetooth
---

\{{APIRef("Bluetooth API")}}\{{SecureContext_Header}}
```

A page that has a banner will usually also have "complementary" page metadata.
For example, a page that has `\{{SeeCompatTable}}` also has the `experimental` status added too (as shown above) to ensure that it has appropriate icons in the sidebar.

> **Note:** Banner macros do not _depend_ on the metadata, but some other macro-inserted content does.
> For example, the `\{{Compat}}` macro depends on the `browser-compat` metadata value.

> ## What banners should be added

Reviewer: ```suggestion
## Banners that must be added manually
```

The [Page type templates](/en-US/docs/MDN/Writing_guidelines/Page_structures/Page_types#templates) include the most important macros.

You need to add the following macros manually:

- `\{{SecureContext_Header}}` — this generates a **Secure context** banner that indicates the technology is only available in a [secure context](/en-US/docs/Web/Security/Secure_Contexts).

> ## What banners are automatically added

Reviewer: ```suggestion
### Banners that must be added automatically
```

The [Page type templates](/en-US/docs/MDN/Writing_guidelines/Page_structures/Page_types#templates) include the most important macros.
In summary:

- `\{{SeeCompatTable}}` — generates a **This is an experimental technology** banner that indicates the technology is [experimental](/en-US/docs/MDN/Writing_guidelines/Experimental_deprecated_obsolete#experimental).
- `\{{Deprecated_Header}}` — generates a **Deprecated** banner that indicates that use of the technology is [discouraged](/en-US/docs/MDN/Writing_guidelines/Experimental_deprecated_obsolete#deprecated).
- `\{{Non-standard_Header}}` — generates a **Non-Standard** banner that indicates that use of the technology is not part of a formal specification, even if it is implemented in multiple browsers.
- `\{{SecureContext_Header}}` — this generates a **Secure context** banner that indicates the technology is only available in a [secure context](/en-US/docs/Web/Security/Secure_Contexts).

The first three technology status related macros in the above list are [automatically added](/en-US/docs/MDN/Writing_guidelines/Page_structures/Technology_status#how_to_add_or_update_technology_statuses) to the content.

> Hence, there is no need to add them manually to the content.

Reviewer 2: ```suggestion
Therefore, the only banner that is still manually added to the content is the **Secure context** banner.
```

This could also be a warning:

````markdown

> **Warning:**
> Do not manually add **\{{SeeCompatTable}}**, **\{{Deprecated_Header}}** or *\{{Non-standard_Header}}** banners...

````

What do you think?

Author: When, in future, a new banner gets introduced the statement won't remain true. Going with the warning.

For example, banners are used to highlight when a particular interface, method or property is deprecated, and should not be used in production code.

This article describes the more important banners and how they are added.

## How to add a banner

Banners are added using macros.
Banner macros should be inserted below the page metadata, alongside the page sidebar macro.
For example,
the `\{{SecureContext_Header}}` macro has been used
to indicate
that the [Bluetooth API](/en-US/docs/Web/API/Bluetooth) is only available in a [secure context](/en-US/docs/Web/Security/Secure_Contexts).

```md
---
title: Bluetooth
slug: Web/API/Bluetooth
page-type: web-api-interface
status:
  - experimental
browser-compat: api.Bluetooth
---

\{{APIRef("Bluetooth API")}}\{{SecureContext_Header}}
```

A page that has a banner will usually also have "complementary" page metadata.
For example, a page that has `\{{Compat}}` should usually have the `browser-compat` property added in front-matter.

> **Note:** Banner macros do not _depend_ on the metadata, but some other macro-inserted content does.
> For example, the `\{{Compat}}` macro depends on the `browser-compat` metadata value.

> ## What banners are/should be added

Reviewer: Fly by.  "Why is "are" needed?
```suggestion
## What banners should be added
```

Author: "are" is for status macros where "are" added automatically by the automation.
"should be" is for other macros that writers are supposed to add.

Reviewer: You might want to consider splitting into two headings for "what you need to do" and "what is done for you".

Reviewer 2: > You might want to consider splitting into two headings for "what you need to do" and "what is done for you".

+1, I think this is the important bit for authors and why I was suggesting to strongly word "do not" to avoid manually editing something that is automated / generated.

Author: done

In summary:

- `\{{SeeCompatTable}}` — generates a **This is an experimental technology** banner that indicates the technology is [experimental](/en-US/docs/MDN/Writing_guidelines/Experimental_deprecated_obsolete#experimental).
- `\{{Deprecated_Header}}` — generates a **Deprecated** banner that indicates that use of the technology is [discouraged](/en-US/docs/MDN/Writing_guidelines/Experimental_deprecated_obsolete#deprecated).
- `\{{Non-standard_Header}}` — generates a **Non-Standard** banner that indicates that use of the technology is not part of a formal specification, even if it is implemented in multiple browsers.
- `\{{SecureContext_Header}}` — this generates a **Secure context** banner that indicates the technology is only available in a [secure context](/en-US/docs/Web/Security/Secure_Contexts).

The first three technology status related macros in the above list are automatically added to the content to match the statuses stored in the `browser-comapt-data` repository.

> **Warning:**
> Do not manually add **\\{{SeeCompatTable}}**, **\\{{Deprecated_Header}}**,
> and **\\{{Non-standard_Header}}** banners to the content.
> Instead,

> > [update the technology status in `browser-compat-data` repository](/en-US/docs/MDN/Writing_guidelines/Page_structures/Technology_status#how_to_add_or_update_technology_statuses).

Reviewer: Is there actual harm in someone doing this? 

I'm always going to add these because I know the state when I start, and before the script is run to get info from BCD. Unless this script is run as part of yari in my client?

Author: > Is there actual harm in someone doing this?

Since beginning of time we do create corresponding PR in BCD to add/update status in BCD. Not doing this was a bad practice as we were getting multiple complains that in page statuses and Compat table statuses at the bottom of the pages weren't matching.

So we made BCD a single source of truth. And started automated syncing content.

> I'm always going to add these because I know the state when I start, and before the script is run to get info from BCD.

If it helps there is no harm in adding these statuses manually while adding feature pages. But don't forget to update BCD repo as well. Not doing so will revert your status changes in content repo (we are not going to do reverse syncing to BCD).

My initial language was in this PR was a suggestion. But later reviewers have changed it to "don't do it".

> Unless this script is run as part of yari in my client?

No point in doing this as you'll have to update BCD repo first and get the PR merged. Then start documenting the new page in content.

Reviewer: OK. I'm probably going to do it anyway on creation, because it is better, but as per https://github.com/mdn/content/pull/31941#discussion_r1470601136 if this is the consensus no worries.

> Banners are added to some pages, in particular API reference, in order to highlight important factors that will affect how the described content is used.

Reviewer: ```suggestion
Banners are displayed in some pages, in particular API reference, in order to highlight important factors that will affect how the described content is used.
```

> For example, banners are used to highlight when a particular interface, method or property is deprecated, and should not be used in production code.

Reviewer: Maybe something like:

```suggestion
For example, banners are used to highlight when a particular interface, method or property is deprecated, and should not be used in production code, or can only be used in a secure context.

Banners are rendered using macros in the page content.
Some banner macros are added automatically to the page, while others are added manually.
```

Banners are added to some pages, in particular API reference, in order to highlight important factors that will affect how the described content is used.
For example, banners are used to highlight when a particular interface, method or property is deprecated, and should not be used in production code.

This article describes the more important banners and how they are added.

## How to add a banner

> Banners are added using macros. Banner macros should be inserted below the page metadata, alongside the page sidebar macro. For example, in the block below, the `\{{SecureContext_Header}}` macro has been used
> to indicate that the {{domxref("Bluetooth")}} interface is only available in a [secure context](/en-US/docs/Web/Security/Secure_Contexts).

Reviewer: ```suggestion
Banners are added using macros. Banner macros should be inserted below the page metadata, alongside the page sidebar macro. For example, in the block below, the `\{{SecureContext_Header}}` macro has been used to indicate that the {{domxref("Bluetooth")}} interface is only available in a [secure context](/en-US/docs/Web/Security/Secure_Contexts).
```

Banners are added to some pages, in particular API reference, in order to highlight important factors that will affect how the described content is used.
For example, banners are used to highlight when a particular interface, method or property is deprecated, and should not be used in production code.

This article describes the more important banners and how they are added.

## How to add a banner

Banners are added using macros.
Banner macros should be inserted below the page metadata, alongside the page sidebar macro.
For example,
the `\{{SecureContext_Header}}` macro has been used
to indicate

> that the [Bluetooth API](/en-US/docs/Web/API/Bluetooth) is only available in a [secure context](/en-US/docs/Web/Security/Secure_Contexts).

Author: Was trying new formatter.

```md
---
title: Bluetooth
slug: Web/API/Bluetooth
page-type: web-api-interface
status:
  - experimental
browser-compat: api.Bluetooth
---

\{{APIRef("Bluetooth API")}}\{{SecureContext_Header}}
```

A page that has a banner will usually also have "complementary" page metadata.

> For example, a page that has `\{{Compat}}` should usually have the `browser-compat` property added in front-matter.

Author: Since contributors do not have to touch these macros, I thought we could mention them less. I've reverted to the old text with some tweaks.

> **Note:** Banner macros do not _depend_ on the metadata, but some other macro-inserted content does.

> > For example, the `\{{Compat}}` macro depends on the `browser-compat` metadata value.

Author: Reverted above para so this is still in business.

> > For example, the `\{{Compat}}` macro depends on the `browser-compat` metadata value.

Reviewer: I'm unresolving this because IMO this whole page needs work, and in particular this particular example needs to be modified.

The whole point of the PR is that you don't need to manually add `\{{SeeCompatTable}}` and you shouldn't need to mess with the metadata. So I would no reference compat macro in the section on macros you have to add.

The way I would try  do this is add another section above `## How to add a banner` that shows what macros and data might appear in content. So showing them everything they might see. Then below explain how stuff is added - either automatically or not.

## What banners should be added

The [Page type templates](/en-US/docs/MDN/Writing_guidelines/Page_structures/Page_types#templates) include the most important macros.

You need to add the following macros manually:

- `\{{SecureContext_Header}}` — this generates a **Secure context** banner that indicates the technology is only available in a [secure context](/en-US/docs/Web/Security/Secure_Contexts).

## What banners are automatically added

Following macros are automatically added to the content in order to match the statuses stored in the `browser-compat-data` repository:

- `\{{SeeCompatTable}}` — generates a **This is an experimental technology** banner that indicates the technology is [experimental](/en-US/docs/MDN/Writing_guidelines/Experimental_deprecated_obsolete#experimental).
- `\{{Deprecated_Header}}` — generates a **Deprecated** banner that indicates that use of the technology is [discouraged](/en-US/docs/MDN/Writing_guidelines/Experimental_deprecated_obsolete#deprecated).
- `\{{Non-standard_Header}}` — generates a **Non-Standard** banner that indicates that use of the technology is not part of a formal specification, even if it is implemented in multiple browsers.

> > **Warning:**
> > Do not manually add **\\{{SeeCompatTable}}**, **\\{{Deprecated_Header}}**,
> > and **\\{{Non-standard_Header}}** banners to the content.
> > Instead,
> > [update the feature status in `browser-compat-data` repository](/en-US/docs/MDN/Writing_guidelines/Page_structures/Feature_status#how_to_add_or_update_feature_statuses).

Reviewer: @OnkarRuikar Split works much better for me thanks. 

You could now change this:

```suggestion
[Update the feature status in `browser-compat-data` repository](/en-US/docs/MDN/Writing_guidelines/Page_structures/Feature_status#how_to_add_or_update_feature_statuses) in order to change these values.

> **Note:** While you can manually/update these macros in content, values that don't match the browser compatibility data will be replaced/removed.
```

> **Note:** While you can manually/update these macros in content, values that don't match the browser compatibility data will be replaced/removed.

> **Note:** Pages that have the `\{{SeeCompatTable}}`, `\{{Deprecated_Header}}`, or `\{{Non-standard_Header}}` banners will also have the corresponding `experimental`, `deprecated` and `non-standard` status values in the page metadata.
> The metadata is automatically updated at the same time as the headers.

> > The banner macros do not depend on this status metadata (but may one day be generated from it).

Author: "but may one day be generated from it" do we need this?
Ideally we would love for yari to render all 9 status macro's content directly based on BCD. And remove the 9 macros from the platform.
